k9mail/src/main/java/com/fsck/k9/preferences/Settings.java

```diff
@@ -97,77 +98,87 @@
      * Upgrade settings using the settings structure and/or special upgrade code.
      *
      * @param version
-     *         The content version of the settings in {@code validatedSettings}.
-     * @param upgraders
+     *         The content version of the settings in {@code validatedSettingsMutable}.
+     * @param customUpgraders
      *         A map of {@link SettingsUpgrader}s for nontrivial settings upgrades.
      * @param settings
      *         The structure describing the different settings, possibly containing multiple
      *         versions.
-     * @param validatedSettings
+     * @param validatedSettingsMutable
      *         The settings as returned by {@link Settings#validate(int, Map, Map, boolean)}.
      *         This map is modified and contains the upgraded settings when this method returns.
      *
      * @return A set of setting names that were removed during the upgrade process or {@code null}
      *         if none were removed.
      */
-    public static Set<String> upgrade(int version, Map<Integer, SettingsUpgrader> upgraders,
+    public static Set<String> upgrade(int version, Map<Integer, SettingsUpgrader> customUpgraders,
             Map<String, TreeMap<Integer, SettingsDescription>> settings,
-            Map<String, Object> validatedSettings) {
-
-        Map<String, Object> upgradedSettings = validatedSettings;
+            Map<String, Object> validatedSettingsMutable) {
         Set<String> deletedSettings = null;
 
         for (int toVersion = version + 1; toVersion <= VERSION; toVersion++) {
-
-            // Check if there's an SettingsUpgrader for that version
-            SettingsUpgrader upgrader = upgraders.get(toVersion);
-            if (upgrader != null) {
-                deletedSettings = upgrader.upgrade(upgradedSettings);
+            if (customUpgraders.containsKey(toVersion)) {
+                SettingsUpgrader upgrader = customUpgraders.get(toVersion);
+                deletedSettings = upgrader.upgrade(validatedSettingsMutable);
```

Comment:
I didn't change semantics here, but it seems odd to me: is keeping only the set of `deletedSettings`  of the last upgrader really what we want to do here?